edx_prefectutils/amplitude.py

```diff
@@ -0,0 +1,347 @@
+
+    Returns:
+        iterable of iterable of iterable of events.
+    """
+    accumulated_events = LazyResultFetcher(cursor, fetch_size=fetch_size)
+    last_idx_for_worker = None
+    while True:
+        if accumulated_events.actual_length() <= batch_size:
+            # All we have left is one or more segments that completely fit into one batch, so send it all!
+            last_idx_for_worker = len(accumulated_events)
+        else:
+            if accumulated_events[0].user_device == accumulated_events[batch_size - 1].user_device:
+                # We have a segment which is larger than a single bucket.  We should find the last event for that
+                # segment and send the entire segment to a worker.
+                target_user_device = accumulated_events[0].user_device
+                idx = batch_size - 1
+                while True:
+                    # Is there a more efficient way than indexing one event at a time?  Yes, but simple iteration is not
+                    # our performance bottleneck, and anything more complicated can introduce bugs.
+                    idx += 1
+                    try:
+                        if accumulated_events[idx].user_device != target_user_device:
+                            last_idx_for_worker = idx
+                            break
+                    except IndexError:
+                        last_idx_for_worker = idx
+                        break
+            else:
+                # One or more segments completely fit into one batch, so lets consolidate them and send them.  Exclude
+                # the last segment so that we send no more than one-batch-worth of events by finding the first event for
+                # the last segment and using that as a cutoff:
+                cut_user_device = accumulated_events[batch_size - 1].user_device
+                idx = batch_size - 1
+                while True:
+                    idx -= 1
+                    if accumulated_events[idx].user_device != cut_user_device:
+                        last_idx_for_worker = idx + 1
+                        break
+
+        # Now we have `last_idx_for_worker` set to the point where we want to slice off events to send to a worker
+        # thread.  Also, it is important that we delete that slice from the accumulated_events so that the python
+        # garbage collector will free all that memory as soon as the corresponding worker thread returns.
+        events_for_next_worker = accumulated_events[:last_idx_for_worker]
+        del accumulated_events[:last_idx_for_worker]
+
+        # Finally, batch the events such that each batch is small enough to fit inside a single Amplitude API call.
+        yield batch_generator(events_for_next_worker, batch_size)
+
+
+def _get_old_high_watermark(sf_credentials: SFCredentials, sf_role: str):
+    """
+    Get the amplitude event ID of the event immediately following the very last event sent to Amplitude's API.
```

Comment:
Ooh, that's confusing.   Almost everywhere else we are using watermarks to indicate the last value loaded, not the next value not loaded.   I'd really encourage us to do the same here.

Reply:
hmm...that does make sense given that sometimes our watermarks are timestamp values which cannot really be "incremented".  I'll have to followup on this.